lib/auth/tls_test.go

```diff
@@ -3528,6 +3516,478 @@ func TestNetworkRestrictions(t *testing.T) {
+	}{
+		{
+			name:     "success",
+			identity: TestUser(privilegedUser.GetName()),
+			token: mustNewToken(
+				t,
+				"success",
+				types.SystemRoles{types.RoleNode, types.RoleTrustedCluster},
+				time.Time{},
+			),
+			requireError:        require.NoError,
+			requireTokenCreated: true,
+			requireEventEmitted: true,
+		},
+		{
+			name:     "already exists",
+			identity: TestUser(privilegedUser.GetName()),
+			token: mustNewToken(
+				t,
+				alreadyExistsToken.GetName(),
+				// These roles differ from the roles on the already existing
+				// token.
+				types.SystemRoles{types.RoleNode, types.RoleTrustedCluster},
+				time.Time{},
+			),
+			requireEventEmitted: true,
+			requireTokenCreated: true,
+			requireError:        require.NoError,
+		},
+		{
+			name:     "access denied",
+			identity: TestNop(),
+			token: mustNewToken(
+				t, "access denied", types.SystemRoles{types.RoleNode}, time.Time{},
+			),
+			requireError: requireAccessDenied,
+		},
+		{
+			name:         "invalid token",
+			identity:     TestUser(privilegedUser.GetName()),
+			token:        &types.ProvisionTokenV2{},
+			requireError: requireBadParameter,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client, err := ac.server.NewClient(tt.identity)
+			require.NoError(t, err)
+
+			mockEmitter.Reset()
+			err = client.UpsertToken(ctx, tt.token)
+			tt.requireError(t, err)
+
+			if tt.requireEventEmitted {
+				lastEvent := mockEmitter.LastEvent()
+				require.NotNil(t, lastEvent)
+				require.Equal(
+					t,
+					events.TrustedClusterTokenCreateEvent,
+					lastEvent.GetType(),
+				)
+			}
+			if tt.requireTokenCreated {
+				token, err := ac.server.Auth().GetToken(ctx, tt.token.GetName())
+				require.NoError(t, err)
+				require.Empty(t, cmp.Diff(
+					tt.token,
+					token,
+					cmpopts.IgnoreFields(types.Metadata{}, "ID"),
+				))
+			}
+		})
+	}
+}
+
+// TestGetTokens tests the GetTokens RPC end-to-end
+func TestGetTokens(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	ac := setupAuthContext(ctx, t)
+
+	// Create a user with the least privilege access to call this RPC.
+	privilegedUser, _, err := CreateUserAndRole(
+		ac.server.Auth(), "token-reader", nil, []types.Rule{
+			{
+				Resources: []string{types.KindToken},
+				Verbs:     []string{types.VerbRead, types.VerbList},
+			},
+		},
+	)
+	require.NoError(t, err)
+
+	// Create Provision token
+	pt := mustNewToken(t, "example-token", types.SystemRoles{types.RoleNode}, time.Time{})
+	require.NoError(t, ac.server.Auth().CreateToken(ctx, pt))
+	// Create static token
+	st, err := types.NewStaticTokens(types.StaticTokensSpecV2{
+		StaticTokens: []types.ProvisionTokenV1{
+			{
+				Roles: types.SystemRoles{types.RoleProxy},
+				Token: "example-static-token",
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.NoError(t, ac.server.Auth().SetStaticTokens(st))
+
+	expectTokens := append([]types.ProvisionToken{pt}, st.GetStaticTokens()...)
```

Comment:
Test with >1 created token, just in case?

Reply:
cbec5b0bde